### PR TITLE
[cuda rng] Making offset calculation independent of device properties

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98988
* #98965

### CODE REVIEW

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 1]
```diff
@@ -51,10 +51,31 @@ std::tuple<uint64_t, dim3, dim3> calc_execution_policy(int64_t total_elements) {
...
+  // number_of_sm = static_cast<uint32_t>(at::cuda::getCurrentDeviceProperties()->multiProcessorCount);
+  //
+  // However, having the offset dependent on cuda device properties makes it
+  // harder/hacky to support functionalization of RNG ops. So we have chosen
+  // lower bounds for the two device properties by looking at the recent GPUs.
+  // Lower bounds ensure that we move the offsets more aggressively, ensuring
+  // that random numbers are not repeated.
+
+
+  // Newer GPUs have 2048 threads per SM, but older ones have 1536. We chose
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DistributionTemplates.h:69]: it's not newer/older, there's a weird thing going with oscillating threads. 
You could also mention that philox sequence length is `2**64` so we are unlikely to run out, and can advance liberally.

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 2]
```diff
@@ -51,10 +51,31 @@ std::tuple<uint64_t, dim3, dim3> calc_execution_policy(int64_t total_elements) {
...
+
+
+  // Number of SMs have been around 1536 and 2048. Choosing 1536 for safety.
+  uint32_t max_threads_per_sm = 1536;
+  // T4 has 40, V100 has 80 and A100 has 108 SMs. To get the lower bound, we
+  // assume MIG on a A100 GPU, i.e. 108/8, and round it a nicer even number 12.
+  uint32_t number_of_sm = 12;
+
+  uint32_t blocks_per_sm = max_threads_per_sm / block_size;
+  grid.x = std::min(number_of_sm * blocks_per_sm, grid.x);
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply aten/src/ATen/native/cuda/DistributionTemplates.h:77]: Good that this is already reverted then :) I will separate out the offset calculation and grid size in next PR.

**Pull request closed without merging**